Upgrade JavaScript libraries in Web Apps

orangejenny opened the pull request:
https://trello.com/c/bnB4soOX/71-marionette-jquery-library-upgrades

This upgrades Marionette from 2.4.5 to 4.1.2 and jQuery from 2.2.4 to 3.5.1. This brings both libraries up to date and means we're no longer using two different versions of jQuery.

There are a lot of commits but each one is pretty small and single-purpose. Links to release notes are in commit comments.

The biggest change is converting all Marionette modules to [hqDefine](https://github.com/dimagi/commcare-hq/blob/master/docs/js-guide/code-organization.md#hqdefine). This PR doesn't migrate web apps to RequireJS, but it gets us a lot closer to that goal.

The second biggest change is that Marionette 3 dropped its `CompositeView` class which necessitated reworking some views...and then that functionality was brought back and added to `CollectionView` in Marionette 4, so there's a series of commits that basically revert the earlier ones. I didn't clean up this git history because it's been handy to be able to jump around in this branch and still have the code functional.

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-1802)

orangejenny committed:
Replaced region attribute access with getRegion

https://marionettejs.com/docs/v3.0.0/upgrade.html#regions

orangejenny committed:
Switched ItemView to LayoutView

https://marionettejs.com/docs/v3.0.0/upgrade.html#removing-layoutview-and-itemview

This is recommended by the migration guide as a backwards-compatible step towards
combining both ItemView and LayoutView into View.

orangejenny committed:
Replaced CompositeView in SettingsView

https://marionettejs.com/docs/v3.0.0/upgrade.html#removing-compositeview

orangejenny committed:
Replaced CompositeView in RestoreAsView

orangejenny committed:
Replaced CompositeView in GridView

orangejenny committed:
Replaced CompositeView in SessionListView (incomplete forms)

orangejenny committed:
Replaced CompositeView in DetailView and DetailTabView

orangejenny committed:
Migrated HQ.Events to hqDefine

https://marionettejs.com/docs/v3.0.0/upgrade.html#modules

orangejenny committed:
Migrated Utils to hqDefine

orangejenny committed:
Removed PhoneNavigation

orangejenny committed:
Updated to 3.0.0

orangejenny committed:
Renamed templateHelpers => templateContext

https://marionettejs.com/docs/v3.0.0/upgrade.html#templates

orangejenny committed:
Replaced LayoutView with View

https://marionettejs.com/docs/v3.0.0/upgrade.html#removing-layoutview-and-itemview

orangejenny committed:
Renamed getChildView => childView

getChildView was removed from CollectionView

https://github.com/marionettejs/marionette-v3-compat#available-patches

orangejenny committed:
Updated table view layouts

https://marionettejs.com/docs/v3.0.0/marionette.collectionview.html#tables-using-marionette-3

orangejenny committed:
Replaced CompositeView in QueryListView (case claim)

orangejenny committed:
Replaced CompositeView in MenuListView

orangejenny committed:
Replaced CompositeView in BreadcrumbView

This one is convoluted. The desire for semantic markup - a single list
containing the static home link plus all subsequent links - is at odds
with Backbone wrapping each view in an element. Ideally there would be
one <ol> with a static home icon and then the CollectionView would
render the remaining list items.

orangejenny committed:
Replaced CompositeView in CaseListView

orangejenny committed:
Migrated Apps.Collections.App to hqDefine

orangejenny committed:
Migrated Apps.Models.App to hqDefine

orangejenny committed:
Migrated Apps.Views to hqDefine

orangejenny committed:
Migrated Apps.Controller to hqDefine

orangejenny committed:
Migrated Apps.API to hqDefine

orangejenny committed:
Migrated Users.Collections.User to hqDefine

orangejenny committed:
Migrated Users.Models to hqDefine

orangejenny committed:
Migrated Users.Views to hqDefine

orangejenny committed:
Migrated Users.Controller to hqDefine

orangejenny committed:
Migrated Sessions namespace to hqDefine

orangejenny committed:
Migrated SessionNavigate.SessionList namespace to hqDefine

orangejenny committed:
Migrated remaining SessionNavigate namespace to hqDefine

orangejenny committed:
Removed Models.MenuSelect

orangejenny committed:
Migrated Menus.Collections to hqDefine

orangejenny committed:
Migrated Menus.Views.QueryListView to hqDefine

orangejenny committed:
Migrated Menus.Views to hqDefine

orangejenny committed:
Migrated Menus.API to hqDefine

orangejenny committed:
Migrated Layout.Views to hqDefine

orangejenny committed:
Split menus/util.js out as a separate file

orangejenny committed:
Migrated Menus.Util to hqDefine (ignore whitespace)

orangejenny committed:
Migrated Menus.Controller to hqDefine (ignore whitespace)

orangejenny committed:
Switched request/reqres, from Wreqr, out for Backbone.Radio

https://github.com/marionettejs/marionette-v3-compat#known-issues

orangejenny committed:
Updated args to start handler

Didn't find this documented, but this handler now seems to receive the object itself as the first argument.

orangejenny committed:
Upgraded to 3.5.1 and fixed tests

3.3.0's change to "getRegion will now render the region's view if it is currently not rendered"
breaks tests that call `show` on a region.

https://github.com/marionettejs/backbone.marionette/blob/master/changelog.md#v330-view-commit-logs

orangejenny committed:
Upgraded to 4.1.2 and added separate AppRouter

https://marionettejs.com/docs/v4.0.0/upgrade-v3-v4.html#approuter-was-removed

orangejenny committed:
Replaced selectors with template functions

Didn't find this explicitly documented in the changelog, but Marionette now
expects templates to be template functions and does not accept selectors.

orangejenny committed:
Switched GridView back to old CompositeView style

https://marionettejs.com/docs/v4.0.0/upgrade-v3-v4.html#compositeview-was-removed

These docs say v4 CollectionView supports the same behavior as v2 CompositeView,
but earlier commits already replaced our v2 CompositeViews for the v3 upgrade.
This reverts to the previous code, jsut replacing the CompositeView with a CollectionView.

orangejenny committed:
Migrated formplayer/constants.js to hqDefine (ignore whitespace)

orangejenny committed:
Added gettext

orangejenny committed:
Removed jQuery 2.2.4

orangejenny committed:
Replaced removeAttr with prop(..., false)

https://jquery.com/upgrade-guide/3.0/#breaking-change-removeattr-no-longer-sets-properties-to-false

orangejenny committed:
Fixed jQuery error on '#' selector

https://jquery.com/upgrade-guide/3.0/#breaking-change-jquery-quot-quot-and-find-quot-quot-are-invalid-syntax

orangejenny committed:
Migrated formplayer/app.js to hqDefine

orangejenny committed:
Switched BreadcrumbView back to old CompositeView style

orangejenny committed:
Switched SessionListView back to old CompositeView style

orangejenny committed:
Switched QueryListView back to old CompositeView style

orangejenny committed:
Switched SettingsView back to old CompositeView style

orangejenny committed:
Switched RestoreAsView back to old CompositeView style

orangejenny committed:
Updated Save button UI in SettingsView

The icon is non-existent, and on Web Apps there's nothing to save since the settings
are all actions with their own buttons (clear user data and break locks).

orangejenny committed:
Switched MenuListView back to old CompositeView style

orangejenny committed:
Switched CaseListView back to old CompositeView style

orangejenny committed:
Switched DetailListView and DetailTabView back to old CompositeView style

stephherbers reviewed (approved):
Let's see what QA has to say

orangejenny committed:
Removed Array.prototype.indexOf polyfill

Browser compatibility is good now:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf#Browser_compatibility

orangejenny committed:
Extracted TaskQueue to its own file

orangejenny committed:
Encapsulated WebFormSession

orangejenny committed:
Encapsulated Util

orangejenny committed:
Extracted cloudcare/js/formplayer/errors

orangejenny committed:
Extracted cloudcare/js/form_entry/const

orangejenny committed:
Extracted cloudcare/js/formplayer/utils

orangejenny committed:
Migrated entrycontrols_full.js to hqDefine

orangejenny committed:
Acquiesced to Stickler

orangejenny committed:
Migrated formplayer-inline.js to hqDefine

orangejenny committed:
Merge pull request #28448 from dimagi/jls/more-web-apps-hqdefine

Migrated more of Web Apps to hqDefine

orangejenny committed:
Merge branch 'master' into jls/marionette-3

orangejenny committed:
Fixed bad merge

orangejenny committed:
Fixed name of errors.js

orangejenny committed:
Removed restore for spy that was deleted

orangejenny committed:
Reduced lint errors

orangejenny committed:
Reduced lint errors

orangejenny committed:
Reduced lint errors

orangejenny committed:
Moved initial page data access to happen after page is loaded

orangejenny committed:
Switched CaseListView.ui.actionButton to use a class since there are two elements

orangejenny committed:
Added translation

orangejenny committed:
Fixed case claim not working in app preview

Now that templates are grabbed when js loads, the templates have to be included before the js.

orangejenny committed:
Fixed circular dependency in cloudcare/js/util

orangejenny committed:
Merge branch 'master' into jls/marionette-3

orangejenny committed:
Fixed hqRequire parameter

orangejenny committed:
Fixed hqRequire call to make sense

orangejenny committed:
Merge branch 'master' into jls/marionette-3
